envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt

```diff
@@ -105,21 +115,52 @@ class EnvoyListenersFactory(
                                         .setAddress(listenersConfig.ingressHost)
                         )
                 )
-                .addFilterChains(createIngressFilterChain(group, listenersConfig, globalSnapshot))
-                .build()
+
+        listOfNotNull(securedIngressChain, insecureIngressChain).forEach {
+            listener.addFilterChains(it.build())
+        }
+
+        return listener.build()
     }
 
     private fun createIngressFilterChain(
         group: Group,
         listenersConfig: ListenersConfig,
-        globalSnapshot: GlobalSnapshot
-    ): FilterChain {
-
-        val filterChain = FilterChain.newBuilder()
+        globalSnapshot: GlobalSnapshot,
+        transportProtocol: String = "raw_buffer"
+    ): FilterChain.Builder {
+        return FilterChain.newBuilder()
+                .setFilterChainMatch(FilterChainMatch.newBuilder().setTransportProtocol(transportProtocol))
                 .addFilters(createIngressFilter(group, listenersConfig, globalSnapshot))
+    }
 
-        return filterChain
+    private fun createSecuredIngressFilterChain(
+        group: Group,
+        listenersConfig: ListenersConfig,
+        globalSnapshot: GlobalSnapshot
+    ): FilterChain.Builder {
+        val filterChain = createIngressFilterChain(group, listenersConfig, globalSnapshot, "tls")
+        val tlsAuthenticationProperties = snapshotProperties.incomingPermissions.tlsAuthentication
+
+        val downstreamTlsContext = DownstreamTlsContext.newBuilder()
+                .setRequireClientCertificate(BoolValue.of(true)) // if false, will not validate cert
+                .setCommonTlsContext(CommonTlsContext.newBuilder()
+                        .addTlsCertificateSdsSecretConfigs(SdsSecretConfig.newBuilder()
+                                .setName(tlsAuthenticationProperties.tlsCertificateConfigName).build()
+                        )
+                        // .addAllAlpnProtocols(listOf("h2,http/1.1"))
+                        .setValidationContextSdsSecretConfig(SdsSecretConfig.newBuilder()
+                                .setName(tlsAuthenticationProperties.validationContextConfigName)
+                                .build()
+                        )
+                )
+        filterChain.setTransportSocket(TransportSocket.newBuilder()
+                .setName("envoy.transport_sockets.tls")
+                .setTypedConfig(ProtobufAny.pack(downstreamTlsContext.build()))
                 .build()
```

Comment:
I think the whole TransportSocket can be a single instance cached as class field.